{% load humanize %}
{% load summary %}
<h2 id="mlst">MLST</h2>
<p>
Multi-Locus sequence types (MLST) are determined in two ways. Processed reads
are mapped to the 7 loci using SRST2. Gene sequences for each loci are also
BLASTed against the assembled contigs.
</p>
<h5>SRST2 Results</h5>
<div class="row">
    <div class="col-lg-6 col-sm-12">
        <table class="table table-striped">
            <thead>
                <tr>
                    <td>Locus</td>
                    <td>Allele<sup>1</sup></td>
                </tr>
            </thead>
            <tbody>
                {% for k, v in srst2_hits.items %}
                    <tr>
                        <td><em>{{ k }}</em></td><td>{{ v }}</td>
                    </tr>
                {% endfor %}
                <tr>
                    <td class="footnote" colspan=2>
                        1: '*' means the best allele had at least 1 mismatch.
                           '?' means the best allele has areas of low coverage.
                           '-' means no allele could be assigned.
                    </td>
                </tr>
                <tr><td></td><td></td></tr>
                <tr>
                    <td class="text-center" colspan=2><strong>SRST2 Statistics<sup>2</sup></strong></td>
                </tr>
                {% for k, v in srst2_stats.items %}
                    <tr>
                        <td><em>{{ k }}</em></td><td>{{ v }}</td>
                    </tr>
                {% endfor %}
                    <td class="footnote" colspan=2>
        2: Brief note about SRST2 statistics.
        <em>Mismatches</em> are the total number of mismatches for each loci (lower is better).
        <em>Depth</em> is the mean coverage across each loci (higher is better).
        <em>maxMAF</em> is the highest minor allele frequency across each loci (lower is better).
        For more detailed information please see <a target="_blank" href="https://github.com/katholt/srst2#mlst-results">SRST2's documentation</a>.
                    </td>
            </tbody>
        </table>
    </div>
</div>

<h5>BLAST Results</h5>
<div class="row">
    <div class="col-lg-8 col-sm-12">
        <table class="table table-responsive table-striped text-center">
            <thead>
                <tr>
                    <td>Locus</td>
                    <td>Allele</td>
                    <td>E-Value</td>
                    <td>Bitscore</td>
                    <td>Alignment Coverage (%)</td>
                    <td>Percent Identical</td>
                    <td>Gaps</td>
                    <td>Mismatches</td>
                </tr>
            </thead>
            <tbody>
            {% for row in blast %}
                <tr>
                    {% for col in row %}
                        {% if forloop.first %}
                        <td><em>{{ col }}</em>.</td>
                        {% else %}
                        <td>{{ col }}</td>
                        {% endif %}
                    {% endfor %}
                </tr>
            {% endfor %}
            </tbody>
        </table>
    </div>
</div>
